### PR TITLE
Zsh completion

### PR DESCRIPTION
**Pull request by FedericoCeratto**
Add --cpu --gc and --os to zsh completion.

It would be nice to have bash completion as well.

(ping @arnetheduck @edubart @PMunch @juancarlospaco FYI)

### CODE REVIEW

File: tools/nim.zsh-completion [hunk 1]
```diff
@@ -68,6 +68,71 @@ _nim() {
...
+    '*--cpu=alpha[compile for Alpha architecture]' \
+    '*--cpu=amd64[compile for amd64/Intel architecture]' \
+    '*--cpu=arm[compile for ARM architecture]' \
+    '*--cpu=arm64[compile for ARM64 architecture]' \
+    '*--cpu=avr[compile for AVR architecture]' \
+    '*--cpu=esp[compile for ESP architecture]' \
+    '*--cpu=hppa[compile for HPPA architecture]' \
+    '*--cpu=i386[compile for i386 architecture]' \
+    '*--cpu=ia64[compile for ia64 architecture]' \
+    '*--cpu=js[compile to javascript]' \
```
Comment (juancarlospaco) [inline tools/nim.zsh-completion]:
```suggestion
    '*--cpu=js[compile to JavaScript]' \
```
is a name needs capital.

File: tools/nim.zsh-completion [hunk 2]
```diff
@@ -68,6 +68,71 @@ _nim() {
...
+    '*--cpu=sparc64[compile for SPARC64 architecture]' \
+    '*--cpu=vm[compile for Nim VM]' \
+    '*--cpu=wasm32[compile to WASM 32]' \
+    '*--gc=refc[use reference counting garbage collection]' \
+    '*--gc=arc[use ARC garbage collection]' \
+    '*--gc=orc[use ORC garbage collection]' \
+    '*--gc=markAndSweep[use mark-and-sweep garbage collection]' \
+    '*--gc=boehm[use Boehm<F2> garbage collection]' \
+    '*--gc=go[use Go garbage collection]' \
+    '*--gc=regions[use region-based garbage collection]' \
```
Comment (juancarlospaco) [inline tools/nim.zsh-completion]: Remove the word "use" on all of them, and capitalize.

Replace "garbage collector" for "memory management".

File: tools/nim.zsh-completion [hunk 3]
```diff
@@ -68,6 +68,71 @@ _nim() {
...
+    '*--cpu=riscv64[compile for RISC-V 64 architecture]' \
+    '*--cpu=sparc[compile for SPARC architecture]' \
+    '*--cpu=sparc64[compile for SPARC64 architecture]' \
+    '*--cpu=vm[compile for Nim VM]' \
+    '*--cpu=wasm32[compile to WASM 32]' \
+    '*--gc=refc[use reference counting garbage collection]' \
+    '*--gc=arc[use ARC garbage collection]' \
+    '*--gc=orc[use ORC garbage collection]' \
+    '*--gc=markAndSweep[use mark-and-sweep garbage collection]' \
+    '*--gc=boehm[use Boehm<F2> garbage collection]' \
```
Comment (juancarlospaco) [inline tools/nim.zsh-completion]: `<F2>`  ??

File: tools/nim.zsh-completion [hunk 4]
```diff
@@ -68,6 +68,71 @@ _nim() {
...
+    '*--cpu=mips64el[compile for MIPS64 EL architecture]' \
+    '*--cpu=msp430[compile for msp430 architecture]' \
+    '*--cpu=nimvm[compile for Nim VM]' \
+    '*--cpu=powerpc[compile for PowerPC architecture]' \
+    '*--cpu=powerpc64[compile for PowerPC64 architecture]' \
+    '*--cpu=powerpc64el[compile for PowerPC64 EL architecture]' \
+    '*--cpu=riscv32[compile for RISC-V 32 architecture]' \
+    '*--cpu=riscv64[compile for RISC-V 64 architecture]' \
+    '*--cpu=sparc[compile for SPARC architecture]' \
+    '*--cpu=sparc64[compile for SPARC64 architecture]' \
```
Comment (juancarlospaco) [inline tools/nim.zsh-completion:94]: Remove the words "compile for " and capitalize IMHO.

File: tools/nim.zsh-completion [hunk 5]
```diff
@@ -68,6 +68,71 @@ _nim() {
...
+    '*--os=NintendoSwitch=[compile for NintendoSwitch]' \
+    '*--os=OS2=[compile for OS2]' \
+    '*--os=OpenBSD=[compile for OpenBSD]' \
+    '*--os=PalmOS=[compile for PalmOS]' \
+    '*--os=QNX=[compile for QNX]' \
+    '*--os=SkyOS=[compile for SkyOS]' \
+    '*--os=Solaris=[compile for Solaris]' \
+    '*--os=VxWorks=[compile for VxWorks]' \
+    '*--os=Windows=[compile for Windows]' \
+    '*--os=JS=[generate javascript]' \
```
Comment (juancarlospaco) [inline tools/nim.zsh-completion]:
```suggestion
    '*--os=JS=[generate JavaScript]' \
```

File: tools/nim.zsh-completion [hunk 6]
```diff
@@ -68,6 +68,71 @@ _nim() {
...
+    '*--os=iOS=[compile for iOS]' \
+    '*--os=Irix=[compile for Irix]' \
+    '*--os=Linux=[compile for Linux]' \
+    '*--os=MacOS=[compile for MacOS]' \
+    '*--os=MacOSX=[compile for MacOSX]' \
+    '*--os=MorphOS=[compile for MorphOS]' \
+    '*--os=NetBSD=[compile for NetBSD]' \
+    '*--os=Netware=[compile for Netware]' \
+    '*--os=NimVM=[compile for NimVM]' \
+    '*--os=NintendoSwitch=[compile for NintendoSwitch]' \
```
Comment (juancarlospaco) [inline tools/nim.zsh-completion]:
```suggestion
    '*--os=NintendoSwitch=[compile for Nintendo Switch]' \
```

File: tools/nim.zsh-completion [hunk 7]
```diff
@@ -68,6 +68,71 @@ _nim() {
...
+    '*--gc=regions[use region-based garbage collection]' \
+    '*--gc=none[disable garbage collection]' \
+    '*--os=Standalone=[generate a stand-alone executable]' \
+    '*--os=AIX=[compile for AIX]' \
+    '*--os=Amiga=[compile for Amiga OS]' \
+    '*--os=Android=[compile for Android]' \
+    '*--os=Any=[compile for any OS]' \
+    '*--os=Atari=[compile for Atari]' \
+    '*--os=DOS=[compile for DOS]' \
+    '*--os=DragonFly=[compile for DragonFly]' \
```
Comment (juancarlospaco) [inline tools/nim.zsh-completion]:
```suggestion
    '*--os=DragonFly=[compile for DragonFlyBSD]' \
```

File: tools/nim.zsh-completion [hunk 8]
```diff
@@ -68,6 +68,71 @@ _nim() {
     '*--app=gui[generate a GUI app]' \
     '*--app=lib[generate a dynamic library]' \
     '*--app=staticlib[generate a static library]' \
+    '*--cpu=alpha[compile for Alpha architecture]' \
+    '*--cpu=amd64[compile for amd64/Intel architecture]' \
```
Comment (juancarlospaco) [inline tools/nim.zsh-completion]:
```suggestion
    '*--cpu=amd64[compile for x86_64 architecture]' \
```

**Comment by timotheecour** [comment]
this is pretty much guaranteed to go out of sync, and violates DRY (EDIT: making the current situation "worse" now you have to edit things in more than 1 place).

Instead we should auto generate advopt.nim, tools/nim.bash-completion, and tools/nim.zsh-completion from a single source; there are many ways for example:
`nim r tools/gencmds.nim`
generates
advopt.nim, tools/nim.bash-completion, and tools/nim.zsh-completion

EDIT: it's also incomplete, even after this PR

**Comment by FedericoCeratto** [comment]
@timotheecour the completion script is being used currently and has been there for years. A completion generator could go in a future PR as it's a different change than the current PR. Unless you want to create a completion generator right now.

File: tools/nim.zsh-completion [hunk 9]
```diff
@@ -68,6 +68,80 @@ _nim() {
...
+    '*--cpu=sparc64[compile for SPARC64 architecture]' \
+    '*--cpu=vm[compile for Nim VM]' \
+    '*--cpu=wasm32[compile to WASM 32]' \
+    '*--gc=refc[use reference counting garbage collection]' \
+    '*--gc=arc[use ARC garbage collection]' \
+    '*--gc=orc[use ORC garbage collection]' \
+    '*--gc=markAndSweep[use mark-and-sweep garbage collection]' \
+    '*--gc=boehm[use Boehm<F2> garbage collection]' \
+    '*--gc=go[use Go garbage collection]' \
+    '*--gc=regions[use region-based garbage collection]' \
```
Comment (Araq) [inline tools/nim.zsh-completion]:
```suggestion
    '*--gc=regions[use region-based memory management]' \
```

File: tools/nim.zsh-completion [hunk 10]
```diff
@@ -68,6 +68,80 @@ _nim() {
...
+    '*--cpu=riscv64[compile for RISC-V 64 architecture]' \
+    '*--cpu=sparc[compile for SPARC architecture]' \
+    '*--cpu=sparc64[compile for SPARC64 architecture]' \
+    '*--cpu=vm[compile for Nim VM]' \
+    '*--cpu=wasm32[compile to WASM 32]' \
+    '*--gc=refc[use reference counting garbage collection]' \
+    '*--gc=arc[use ARC garbage collection]' \
+    '*--gc=orc[use ORC garbage collection]' \
+    '*--gc=markAndSweep[use mark-and-sweep garbage collection]' \
+    '*--gc=boehm[use Boehm<F2> garbage collection]' \
```
Comment (Araq) [inline tools/nim.zsh-completion]:
```suggestion
    '*--gc=boehm[use Boehm garbage collection]' \
```

**Pull request merged**